File: tests/cuda/python/test_linalg.py
"""Linalg CUDA binding tests."""

import sys
import os
import pytest

_root = os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "..", "..")
sys.path.insert(0, os.path.join(_root, "bindings", "python"))
sys.path.insert(0, os.path.join(_root, "build", "bindings", "python"))

try:
    import insight as ins
    import numpy as np
except ImportError:
    pytest.skip("Insight or NumPy not available", allow_module_level=True)

try:
    ins.init()
    if ins.device_count() == 0:
        pytest.skip("No GPU available", allow_module_level=True)
except Exception:
    pytest.skip("GPU backend not available", allow_module_level=True)

GPU = ins.GPUPlace(0)
CPU = ins.CPUPlace()


def to_gpu(np_arr, dtype=None):
    if dtype is not None:
        np_arr = np_arr.astype(dtype)
    return ins.from_numpy(np_arr).to(GPU)


def to_numpy(gpu_arr):
    return gpu_arr.to(CPU).numpy()


class TestLinalgCUDA:
    def test_matmul(self):
        a_np = np.array([[1, 2], [3, 4]], dtype=np.float64)
        b_np = np.array([[5, 6], [7, 8]], dtype=np.float64)
        result = ins.matmul(to_gpu(a_np), to_gpu(b_np))
        np.testing.assert_allclose(to_numpy(result), a_np @ b_np, rtol=1e-10)

    def test_dot(self):
        a_np = np.array([1, 2, 3], dtype=np.float64)
        b_np = np.array([4, 5, 6], dtype=np.float64)
        result = ins.dot(to_gpu(a_np), to_gpu(b_np))
        expected = np.dot(a_np, b_np)
        assert to_numpy(result).item() == pytest.approx(expected, rel=1e-10)

    def test_outer(self):
        a_np = np.array([1, 2, 3], dtype=np.float64)
        b_np = np.array([4, 5], dtype=np.float64)
        result = ins.outer(to_gpu(a_np), to_gpu(b_np))
        np.testing.assert_allclose(to_numpy(result), np.outer(a_np, b_np), rtol=1e-10)

    def test_norm_vector(self):
        a_np = np.array([3, 4], dtype=np.float64)
        result = ins.norm(to_gpu(a_np))
        assert to_numpy(result).item() == pytest.approx(5.0, rel=1e-5)

    def test_norm_matrix(self):
        a_np = np.array([[1, 2], [3, 4]], dtype=np.float64)
        result = ins.norm(to_gpu(a_np))
        expected = np.linalg.norm(a_np, "fro")
        assert to_numpy(result).item() == pytest.approx(expected, rel=1e-5)

    def test_trace(self):
        a_np = np.array([[1, 2], [3, 4]], dtype=np.float64)
        result = ins.trace(to_gpu(a_np))
        assert to_numpy(result).item() == pytest.approx(5.0, rel=1e-10)

    def test_det(self):
        a_np = np.array([[1, 2], [3, 4]], dtype=np.float64)
        result = ins.det(to_gpu(a_np))
        assert to_numpy(result).item() == pytest.approx(-2.0, rel=1e-5)

    def test_det_3x3(self):
        a_np = np.array([[6, 1, 1], [4, -2, 5], [2, 8, 7]], dtype=np.float64)
        result = ins.det(to_gpu(a_np))
        expected = np.linalg.det(a_np)
        assert to_numpy(result).item() == pytest.approx(expected, rel=1e-5)

    def test_inv(self):
        a_np = np.array([[4, 7], [2, 6]], dtype=np.float64)
        result = ins.inv(to_gpu(a_np))
        np.testing.assert_allclose(to_numpy(result), np.linalg.inv(a_np), rtol=1e-5)

    def test_solve(self):
        a_np = np.array([[3, 1], [1, 2]], dtype=np.float64)
        b_np = np.array([9, 8], dtype=np.float64)
        result = ins.solve(to_gpu(a_np), to_gpu(b_np))
        expected = np.linalg.solve(a_np, b_np)
        np.testing.assert_allclose(to_numpy(result), expected, rtol=1e-5)

    def test_svd(self):
        a_np = np.array([[1, 2], [3, 4], [5, 6]], dtype=np.float64)
        u, s, vt = ins.svd(to_gpu(a_np))
        _, s_np, _ = np.linalg.svd(a_np, full_matrices=False)
        np.testing.assert_allclose(to_numpy(s), s_np, rtol=1e-5)

    def test_eigh(self):
        a_np = np.array([[4, 2], [2, 3]], dtype=np.float64)
        w, v = ins.eigh(to_gpu(a_np))
        w_np, _ = np.linalg.eigh(a_np)
        np.testing.assert_allclose(np.sort(to_numpy(w)), np.sort(w_np), rtol=1e-5)

    def test_cholesky(self):
        a_np = np.array([[4, 2], [2, 3]], dtype=np.float64)
        result = ins.cholesky(to_gpu(a_np))
        expected = np.linalg.cholesky(a_np)
        np.testing.assert_allclose(to_numpy(result), expected, atol=1e-10)

    def test_qr(self):
        a_np = np.array([[1, 2], [3, 4], [5, 6]], dtype=np.float64)
        q, r = ins.qr(to_gpu(a_np))
        _, r_np = np.linalg.qr(a_np)
        np.testing.assert_allclose(to_numpy(r), r_np, atol=1e-5)

    def test_matrix_power(self):
        a_np = np.array([[1, 2], [3, 4]], dtype=np.float64)
        result = ins.matrix_power(to_gpu(a_np), 2)
        expected = np.linalg.matrix_power(a_np, 2)
        np.testing.assert_allclose(to_numpy(result), expected, rtol=1e-5)

    def test_slogdet(self):
        a_np = np.array([[1, 2], [3, 4]], dtype=np.float64)
        sign, logabsdet = ins.slogdet(to_gpu(a_np))
        sign_np, logabsdet_np = np.linalg.slogdet(a_np)
        assert to_numpy(sign).item() == pytest.approx(sign_np, abs=1e-5)
        assert to_numpy(logabsdet).item() == pytest.approx(logabsdet_np, rel=1e-5)

    def test_eigvalsh(self):
        a_np = np.array([[4, 2], [2, 3]], dtype=np.float64)
        w = ins.eigvalsh(to_gpu(a_np))
        w_np = np.linalg.eigvalsh(a_np)
        np.testing.assert_allclose(np.sort(to_numpy(w)), np.sort(w_np), rtol=1e-5)

    def test_pinv(self):
        a_np = np.array([[1, 2], [3, 4], [5, 6]], dtype=np.float64)
        result = ins.pinv(to_gpu(a_np))
        expected = np.linalg.pinv(a_np)
        np.testing.assert_allclose(to_numpy(result), expected, rtol=1e-5)

    def test_lstsq(self):
        a_np = np.array([[1, 1], [1, 2], [1, 3]], dtype=np.float64)
        b_np = np.array([1, 2, 2], dtype=np.float64)
        result = ins.lstsq(to_gpu(a_np), to_gpu(b_np))
        expected, _, _, _ = np.linalg.lstsq(a_np, b_np, rcond=None)
        np.testing.assert_allclose(to_numpy(result), expected, rtol=1e-5)

    def test_cond(self):
        a_np = np.array([[1, 0], [0, 2]], dtype=np.float64)
        result = ins.cond(to_gpu(a_np))
        expected = np.linalg.cond(a_np)
        assert to_numpy(result).item() == pytest.approx(expected, rel=1e-5)


if __name__ == "__main__":
    pytest.main([__file__, "-v"])

File: tests/cuda/julia/test_linalg.jl
# Linalg CUDA binding tests
# Run with:
#   LD_LIBRARY_PATH=build/backends/cpu:build/backends/cuda julia tests/cuda/julia/test_linalg.jl

push!(LOAD_PATH, joinpath(@__DIR__, "..", "..", "..", "bindings", "julia"))
push!(LOAD_PATH, joinpath(@__DIR__, "..", "..", "..", "build", "bindings", "julia"))

using Insight

# Try GPU
try
    Insight.load_backend("gpu")
    Insight.set_device(Insight.GPUPlace(0))
catch e
    println("SKIP: GPU backend not available: $e")
    exit(0)
end

passed = 0
failed = 0

function check(name, cond)
    global passed, failed
    if cond
        passed += 1
    else
        failed += 1
        println("FAIL: $name")
    end
end

println("=== Linalg CUDA ===")

# matmul
a = Insight.to_array(Insight.from_data([1.0, 2.0, 3.0, 4.0]), Insight.GPUPlace(0))
a = Insight.reshape(a, [2, 2])
b = Insight.to_array(Insight.from_data([5.0, 6.0, 7.0, 8.0]), Insight.GPUPlace(0))
b = Insight.reshape(b, [2, 2])
c = Insight.matmul(a, b)
check("matmul", Insight.numel(c) == 4)

# dot
a1 = Insight.to_array(Insight.from_data([1.0, 2.0, 3.0]), Insight.GPUPlace(0))
b1 = Insight.to_array(Insight.from_data([4.0, 5.0, 6.0]), Insight.GPUPlace(0))
d = Insight.dot(a1, b1)
check("dot", Insight.numel(d) == 1)

# outer
o = Insight.outer(a1, b1)
check("outer", Insight.numel(o) == 6)

# norm
v = Insight.to_array(Insight.from_data([3.0, 4.0]), Insight.GPUPlace(0))
n = Insight.norm(v)
check("norm", Insight.numel(n) == 1)

# trace
t = Insight.trace(a)
check("trace", Insight.numel(t) == 1)

# det
try
    local d = Insight.det(a)
    check("det", Insight.numel(d) == 1)
catch e
    println("SKIP: det ($e)")
end

# inv
try
    local a_inv = Insight.inv(a)
    check("inv", Insight.numel(a_inv) == 4)
catch e
    println("SKIP: inv ($e)")
end

# solve
try
    a2 = Insight.to_array(Insight.from_data([3.0, 1.0, 1.0, 2.0]), Insight.GPUPlace(0))
    a2 = Insight.reshape(a2, [2, 2])
    b2 = Insight.to_array(Insight.from_data([9.0, 8.0]), Insight.GPUPlace(0))
    local x = Insight.solve(a2, b2)
    check("solve", Insight.numel(x) == 2)
catch e
    println("SKIP: solve ($e)")
end

# svd
try
    a3 = Insight.to_array(Insight.from_data([1.0, 2.0, 3.0, 4.0, 5.0, 6.0]), Insight.GPUPlace(0))
    a3 = Insight.reshape(a3, [3, 2])
    local u, s, vt = Insight.svd(a3)
    check("svd", Insight.numel(s) == 2)
catch e
    println("SKIP: svd ($e)")
end

# cholesky
try
    a4 = Insight.to_array(Insight.from_data([4.0, 2.0, 2.0, 3.0]), Insight.GPUPlace(0))
    a4 = Insight.reshape(a4, [2, 2])
    local l = Insight.cholesky(a4)
    check("cholesky", Insight.numel(l) == 4)
catch e
    println("SKIP: cholesky ($e)")
end

# matrix_power
try
    local mp = Insight.matrix_power(a, 2)
    check("matrix_power", Insight.numel(mp) == 4)
catch e
    println("SKIP: matrix_power ($e)")
end

# slogdet
try
    local s, ld = Insight.slogdet(a)
    check("slogdet", Insight.numel(s) == 1 && Insight.numel(ld) == 1)
catch e
    println("SKIP: slogdet ($e)")
end

# eigvalsh
try
    local w = Insight.eigvalsh(a)
    check("eigvalsh", Insight.numel(w) == 2)
catch e
    println("SKIP: eigvalsh ($e)")
end

# pinv
try
    a6 = Insight.to_array(Insight.from_data([1.0, 2.0, 3.0, 4.0, 5.0, 6.0]), Insight.GPUPlace(0))
    a6 = Insight.reshape(a6, [3, 2])
    local pi = Insight.pinv(a6)
    check("pinv", Insight.numel(pi) == 6)
catch e
    println("SKIP: pinv ($e)")
end

# lstsq
try
    a7 = Insight.to_array(Insight.from_data([1.0, 1.0, 1.0, 2.0, 1.0, 3.0]), Insight.GPUPlace(0))
    a7 = Insight.reshape(a7, [3, 2])
    b7 = Insight.to_array(Insight.from_data([1.0, 2.0, 2.0]), Insight.GPUPlace(0))
    local x = Insight.lstsq(a7, b7)
    check("lstsq", Insight.numel(x) == 2)
catch e
    println("SKIP: lstsq ($e)")
end

# cond
try
    local c = Insight.cond_fn(a)
    check("cond", Insight.numel(c) == 1)
catch e
    println("SKIP: cond ($e)")
end

# matrix_rank
try
    local r = Insight.matrix_rank(a)
    check("matrix_rank", Insight.numel(r) == 1)
catch e
    println("SKIP: matrix_rank ($e)")
end

# ============================================================================
# Results
# ============================================================================
println("\n" * "="^40)
println("Results: $passed passed, $failed failed")
if failed > 0
    exit(1)
end
